### ISSUE TITLE
System memory leak on cuda GPU backend.

### ISSUE THREAD

**Issue by nttstar** [microsoft/onnxruntime#8147]
**Describe the bug**
System memory keeps increasing while using the CUDA GPU backend.

**Urgency**
very urgent

**System information**
- OS Platform and Distribution : Linux Ubuntu 16.04
- ONNX Runtime installed from (source or binary): pip install onnxruntime-gpu==1.8
- ONNX Runtime version: 1.8
- Python version: 3.7.10
- Visual Studio version (if applicable): No
- GCC/Compiler version (if compiling from source): - 
- CUDA/cuDNN version: 11.1
- GPU model and memory: A30, 24GB

**To Reproduce**

`pip install insightface`

```
import insightface
import numpy as np
from insightface.app import FaceAnalysis

app = FaceAnalysis(allowed_modules=['detection']) #it will download models to `~/.insightface/models/`
app.prepare(ctx_id=0) #use gpu backend
img = np.random.randint(0,255,size=(640,640,3), dtype=np.uint8)
for _ in range(1000000):
     app.get(img)
```

The leak is happening at `https://github.com/deepinsight/insightface/blob/master/python-package/insightface/model_zoo/scrfd.py#L152` which is a straightforward inference code.

**Expected behavior**
system memory cost is stable

**Screenshots**
-

**Additional context**
if we set `app.prepare(ctx_id=-1)` to use CPU backend, there's no such bug.

**Comment by yuslepukhin** [comment]
The code and the leak location does not indicate ORT involvement here. Perhaps, you could supply more info along with the model if possible.

**Comment by nttstar** [comment]
> The code and the leak location does not indicate ORT involvement here. Perhaps, you could supply more info along with the model if possible.

@yuslepukhin Thanks for your attention, I have updated the issue description.

**Comment by yuslepukhin** [comment]
Can you insert gc.collect() right after run and see if anything changes?

**Comment by nttstar** [comment]
> Can you insert gc.collect() right after run and see if anything changes?

@yuslepukhin Nothing changed.

**Comment by yuslepukhin** [comment]
And you are feeding a numpy array returned from cv2.dnn.blobFromImage()?

**Comment by nttstar** [comment]
> And you are feeding a numpy array returned from cv2.dnn.blobFromImage()?

@yuslepukhin Yes.  It's not the key problem, I believe. Ver 1.6 works well for this piece of code.

**Comment by yuslepukhin** [comment]
There is a repro.

**Comment by mrjarhead** [comment]
FYI - Ran into this same memory issue when running CUDA inference for a set of deep learning models (RetinaFace+ArcFace+Age Estimation+Custom Classifier+YoloV4) with ORT 1.8.1 (C# API).  The memory leak takes a number of runs before it really starts to rear it's head.  Originally went down the rabbit hole of tracing any memory leaks from items not being disposed, but it turns out it was the ORT+CUDA version combination (issue does not occur when running the CPU EP).  Upgraded CUDA to 11.4.1 and the latest cuDNN version, and life appears to be good now.  

TLDR;
ONNX Runtime 1.8.1 CPU EP == No Memory Leak
ONNX Runtime 1.8.1 CUDA EP - CUDA 11.1+cuDNN 8.0.4.30 == Memory Leak
ONNX Runtime 1.8.1 CUDA EP - CUDA 11.4.1+cuDNN 8.2.2.26 == No Memory Leak

I have not tested extensively on any other CUDA or ORT builds.  Hope this helps and saves someone time in the future!

**Comment by CarlPoirier** [comment]
I too can confirm this. My environment is Windows Server 2016. Here are the CUDA versions used:

ONNX Runtime 1.9.0 CUDA EP - CUDA 11.1.1+cuDNN 8.0.5.39 == Memory Leak
ONNX Runtime 1.9.0 CUDA EP - CUDA 11.4.3+cuDNN 8.2.4.15 == No Memory Leak

**Comment by LoveU3tHousand2** [comment]
onnxruntime 1.14.1  cuda 11.6 cudnn 8.9.0  Memory Leak. very bad.
